### PR TITLE
makefiles: add FEATURES_OPTIONAL_ANY

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

From looking at  #15836 and #15817, it's clear that some dependencies are better modeled in Kconfig with choices. This PR looks to give a similar option to model this behavior in `make` while the migration is ongoing.

So similar to #13349:

> - `FEATURES_OPTIONAL_ALT` are nice to have `FEATURES` that fulfill a similar
>   function. Alternatives are separated by a pipe (`|`) in order of preference,
>   e.g.: `FEATURES_OPTIONAL_ALT += puf_sram|periph_hwrng` if both are provided
>   then only `puf_sram` will be used.

### Testing procedure

#12166 Shows a test case for this (I want to rebase that one on top of  #15836 and this one)

The save_all_dependencies script should have the same output as before, as this PR does not make use of the feature.

### Issues/PRs references

Taken from #12166
Related to  #15836

### CODE REVIEW

File: makefiles/features_check.inc.mk [hunk 1]
```diff
@@ -1,41 +1,66 @@
...
+FEATURES_OPTIONAL_ONE_OUT_OF := $(call _features_one_out_of,\
+                                  $(FEATURES_OPTIONAL_ALT),\
+                                    $(FEATURES_OPTIONAL) \
+                                    $(FEATURES_USABLE))
+
+FEATURES_OPTIONAL_ONLY := $(sort $(filter-out $(FEATURES_REQUIRED),\
+                            $(FEATURES_OPTIONAL) \
+                            $(FEATURES_OPTIONAL_ONE_OUT_OF)))
+FEATURES_OPTIONAL_USED := $(sort $(filter $(FEATURES_PROVIDED), \
+                            $(FEATURES_OPTIONAL_ONLY)))
```
Comment (fjmolinas) [inline makefiles/features_check.inc.mk:45]: After a comment from Leandro offline, this should probably be:

```suggestion
# FEATURES_OPTIONAL that have not been required
FEATURES_OPTIONAL_ONLY_SO_FAR := $(sort $(filter-out $(FEATURES_REQUIRED),\
                            $(FEATURES_OPTIONAL)))
# Only add FEATURES that are usable
FEATURES_OPTIONAL_USED_SO_FAR := $(sort $(filter $(FEATURES_USABLE) \
                                    $(FEATURES_OPTIONAL_ONLY_SO_FAR)))


# Additionally optional features due to the "one out of" dependencies,
#   For each features list in $(FEATURES_OPTIONAL_ALT):
#     ==> If one (or more) features is already added to OPTIONAL_USED
          then the firs of the used ones is returned
#     ==> If one (or more) features is usable select the first one in list
#     ==> If no feature is usable return the list
FEATURES_OPTIONAL_ONE_OUT_OF := $(call _features_one_out_of,\
                                  $(FEATURES_OPTIONAL_ALT),\
                                    $(FEATURES_OPTIONAL_USED_SO_FAR) \
                                    $(FEATURES_USABLE))

# FEATURES_OPTIONAL that have not been required
FEATURES_OPTIONAL_ONLY := $(sort $(filter-out $(FEATURES_REQUIRED),\
                            $(FEATURES_OPTIONAL_ONE_OUT_OF)))
# Only add FEATURES that are usable
FEATURES_OPTIONAL_USED := $(sort $(filter $(FEATURES_USABLE) \
                                    $(FEATURES_OPTIONAL_ONLY )))
```

This way do not add to `FEATURES_OPTIONAL_USED` non-usable features, and do not add with the `FEATURES_OPTIONAL_ALT` a feature that will be added bt the OPTIONAL mechanism. In other words:

`FEATURES_OPTIONAL` has precedence over `FEATURES_OPTIONAL_ALT`

**Review by leandrolanzieri — COMMENTED** [review]
I quickly tried this on the hello-world example (samr21-xpro), and I am not getting the expected result, the last feature of the list is selected. Am I using it wrongly?

<details><summary>Normal used features</summary>

```
arch_32bit
arch_arm
cpu_core_cortexm
no_idle_thread
periph_gpio
periph_pm
periph_uart
```
</details>

After adding `FEATURES_OPTIONAL_ALT += periph_cpuid|periph_timer|periph_i2c` to the application Makefile:
<details><summary>New used features</summary>

```
arch_32bit
arch_arm
cpu_core_cortexm
no_idle_thread
periph_gpio
periph_i2c
periph_pm
periph_uart
```
</details>

**Comment by fjmolinas** [comment]
@leandrolanzieri I addressed the issue you pointed out in the first fixup e148d7b, the comment I inlined is addressed in 6c9983b

```
FEATURES_OPTIONAL_ALT="periph_cpuid|periph_timer|periph_pwm" make -C examples/hello-world/ info-modules
auto_init
board
core
core_idle_thread
core_init
core_msg
core_panic
cpu
native_drivers
periph
periph_common
periph_cpuid
periph_gpio
periph_gpio_linux
periph_init
periph_init_cpuid
periph_init_gpio
periph_init_gpio_linux
periph_init_pm
periph_init_uart
periph_pm
periph_uart
stdio_native
sys
```

**Comment by fjmolinas** [comment]
Note that in master the issue you pointed out was also present:

```
FEATURES_REQUIRED_ANY="periph_cpuid|periph_timer|periph_pwm" make -C examples/hello-world/ info-modules
auto_init
board
core
core_idle_thread
core_init
core_msg
core_panic
cpu
native_drivers
periph
periph_common
periph_gpio
periph_gpio_linux
periph_init
periph_init_gpio
periph_init_gpio_linux
periph_init_pm
periph_init_timer
periph_init_uart
periph_pm
periph_timer
periph_uart
stdio_native
sys
```

**Comment by fjmolinas** [comment]
> Note that in master the issue you pointed out was also present:
> 

So nice catch!

File: doc/doxygen/src/build-system-basics.md [hunk 2]
```diff
@@ -55,6 +55,11 @@ For a `FEATURE` to be provided by a `board` it must meet 2 criteria, and for
 - `FEATURES_OPTIONAL` are "nice to have" `FEATURES`, not needed but useful. If
   available they are always included.
 
+- `FEATURES_OPTIONAL_ALT` are nice to have `FEATURES` that fullfil a similar
```
Comment (maribu) [inline doc/doxygen/src/build-system-basics.md:58]:
```suggestion
- `FEATURES_OPTIONAL_ALT` are nice to have `FEATURES` that fulfill a similar
```

Also, why not naming it `FEATURES_OPTIONAL_ANY`, as it behaves to `FEATURES_REQUIRED_ANY` the same way as `FEATURES_OPTIONAL` behave to `FEATURES_REQUIRED`?

Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md:58]: Both are correct, `fullfil` is the british english.

Comment (fjmolinas) [reply doc/doxygen/src/build-system-basics.md:58]: The name change makes sense though. Will do.

**Comment by maribu** [comment]
Please squash :-) (Also note that commit message needs updating.)

**Comment by maribu** [comment]
Wait, I found an issue:

```
FEATURES_OPTIONAL_ANY="periph_i2c|periph_gpio" BOARD=samr21-xpro make info-modules -C examples/
hello-world/
[...]
periph_gpio
periph_i2c
[...]
```

The `periph_i2c` shoudln't be pulled in. Same issue with `FEATURES_REQUIRED_ANY`. It does work as expected with `FEATURES_REQUIRED_ANY` in `master`.

**Comment by aabadie** [comment]
I think there's an issue with this PR: the number of jobs on Murdock is suspiciously low. The total run took 20 minutes, compared to the others, some applications are not built anymore.

```
compile job results (0 failed, 27546 passed, 27546 total)
```

Nowadays, there +80k jobs on Murdock.

**Comment by fjmolinas** [comment]
> I think there's an issue with this PR: the number of jobs on Murdock is suspiciously low. The total run took 20 minutes, compared to the others, some applications are not built anymore.
> 
> ```
> compile job results (0 failed, 27546 passed, 27546 total)
> ```
> 
> Nowadays, there +80k jobs on Murdock.

Indeed that is strange, I'll take a look.

**Comment by fjmolinas** [comment]
Hmm In Murdock I see that some jobs did not get built for a bunch of `BOARD`s, `tests/saul` does not get built for any Nucleo. But the `info-board-supported` output matches what is expected. Might it have something to do with the fact that a conflict showed up during the build? I'll rebase and run the dependencies script in any case.

**Comment by fjmolinas** [comment]
After all these bugs were uncovered with the `any` algorithm I've realized that to implement this some question would need answering first;

In what order should a `FEATURES_OPTIONAL_ANY` feature be resolved:
- e.g.: if there is an overlap with `FEATURES_REQUIRED_ANY` I would expect only the overlap to be selected, but that means resolving `OPTIONAL_ANY` after `REQUIRED_ANY`.

But since the dependency process is iterative there are no guarantees of this anyway. So I think this is going to become messy very fast, any I'm not sure its worth the effort anymore. I'll still try to take a look at how it would look as soon as I have some time, but as I said I'm not sure its worth it.

**Comment by maribu** [comment]
I think we should aim for best effort here (that is, after `FEATURES_REQUIRED_ANY` of the same Iteration). Without access to the full dependency graph, we cannot do much bettee anyway IMO.

**Comment by maribu** [comment]
I kind of assumed that this has already landed. @fjmolinas: Will you reboot this, or is it better to get KConfig finished so that we no longer need this?

**Pull request closed without merging**